src/clojure/puppetlabs/ssl_utils/simple.clj

```diff
@@ -44,30 +44,51 @@
    :post [(validity-date-range? %)]}
   (let [now        (time/now)
         not-before (time/minus now (time/days 1))
-        not-after  (time/plus now (time/secs ca-ttl))]
+        not-after  (time/plus now (time/seconds ca-ttl))]
     {:not-before (.toDate not-before)
      :not-after  (.toDate not-after)}))
 
-(def key-length 4096)
+(def default-keylength
+  "The default bit length to use when generating keys.
+   Note that all API functions accept an `options' map which may
+   have a :keylength for specifying this value."
+  4096)
 
 (defn gen-keys
-  [certname]
-  {:pre [(string? certname)]
-   :post [(ssl-keys? %)]}
-  (let [keypair     (ssl-utils/generate-key-pair key-length)]
-    {:public-key (ssl-utils/get-public-key keypair)
-     :private-key (ssl-utils/get-private-key keypair)
-     :x500-name (ssl-utils/cn certname)
-     :certname certname}))
+  "Generate public and private keys and the X500 name for the given `certname'.
+   An optional map may be provided to specify:
+
+   * :keylength  Bit length to use for the public/private keys;
+                 defaults to 4096."
+  ([certname] (gen-keys certname {}))
+  ([certname options]
+   {:pre [(string? certname)
+          (map? options)]
+    :post [(ssl-keys? %)]}
+   (let [keylength (get options :keylength default-keylength)
+         keypair (ssl-utils/generate-key-pair keylength)]
+     {:public-key (ssl-utils/get-public-key keypair)
+      :private-key (ssl-utils/get-private-key keypair)
+      :x500-name (ssl-utils/cn certname)
+      :certname certname})))
 
 (defn gen-cert*
-  [ca-keys host-keys serial]
-  {:pre [(ssl-keys? ca-keys)
-         (ssl-keys? host-keys)
-         (integer? serial)]
-   :post [(ssl-utils/certificate? %)]}
-  (let [validity (cert-validity-dates (* 5 60 60 24 365))]
-    (ssl-utils/sign-certificate
+  "Internal helper function to generate a certificate; see `gen-cert' for the
+   public version of this function.
+   An optional map may be provided to specify:
+
+   * :extensions  List of certificate extensions to include on the certificate;
+                  defaults to []."
+  ([ca-keys host-keys serial] (gen-cert* ca-keys host-keys serial {}))
```

Comment:
Seems like we wouldn't need to support the `options`-less form of calling into this function if we assume it will only be called from the "public" `gen-cert`, which would always call with an `options` map.  Not a big deal, though.